This needs a test in a subprocess. In POSIX the maximum available address space or heap size can be easily controlled. For example:
```
>>> import resource
>>> hard_limit = resource.getrlimit(resource.RLIMIT_AS)[1]
>>> resource.setrlimit(resource.RLIMIT_AS, (128 * (1<<20), hard_limit))
>>> f = open('test.txt', 'w')
>>> f.write('a' * (64 * (1<<20)))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
MemoryError
>>> f.write('a')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
MemoryError
```
It's possible in Windows with kernel Job objects, but Python's standard library doesn't support them, not without ctypes.

--- a/Lib/test/test_io.py
+++ b/Lib/test/test_io.py
@@ -3767,6 +3767,33 @@ class CTextIOWrapperTest(TextIOWrapperTest):
         with self.assertRaises(AttributeError):
             del t._CHUNK_SIZE
 
+    def test_internal_buffer_size(self):
+        # bpo-43260: TextIOWrapper's internal buffer should not store
+        # data larger than chunk size.
+        chunk_size = 8192  # default chunk size, updated later
+
+        class MockIO(self.MockRawIO):
+            def write(self, data):
+                if len(data) > chunk_size:
+                    raise RuntimeError
+                return super().write(data)
+
+        buf = MockIO()
+        t = self.TextIOWrapper(buf, encoding="ascii")
+        chunk_size = t._CHUNK_SIZE
+        t.write("abc")
+        t.write("def")
+        # default chunk size is 8192 bytes so t don't write data to buf.
+        self.assertEqual([], buf._write_stack)
+
+        with self.assertRaises(RuntimeError):
+            t.write("x"*(chunk_size+1))
+
+        self.assertEqual([b"abcdef"], buf._write_stack)
+        t.write("ghi")
+        t.write("x"*chunk_size)
+        self.assertEqual([b"abcdef", b"ghi", b"x"*chunk_size], buf._write_stack)
+
 
 class PyTextIOWrapperTest(TextIOWrapperTest):
     io = pyio

--- a/Modules/_io/textio.c
+++ b/Modules/_io/textio.c
@@ -1684,7 +1684,7 @@ _io_TextIOWrapper_write_impl(textio *self, PyObject *text)
         self->pending_bytes = b;
     }
     else if (self->pending_bytes_count + bytes_len > self->chunk_size) {
-        // Prevent to concatinate more than chunk_size data.
+        // Prevent to concatenate more than chunk_size data.
         if (_textiowrapper_writeflush(self) < 0) {
             Py_DECREF(b);
             return NULL;

